File: src/main/java/com/freya02/bot/versioning/github/UpdateCountdown.java
package com.freya02.bot.versioning.github;

import java.util.concurrent.TimeUnit;

public class UpdateCountdown {
	private final long interval;

	private long lastUpdate = 0;

	public UpdateCountdown(long time, TimeUnit unit) {
		this.interval = unit.toMillis(time);
	}

	public boolean needsUpdate() {
		if ((System.currentTimeMillis() - lastUpdate) > interval) {
			lastUpdate = System.currentTimeMillis();

			return true;
		} else {
			return false;
		}
	}
}

File: doxxy-bot/src/main/kotlin/dev/freya02/doxxy/bot/versioning/github/UpdateCountdown.kt
package dev.freya02.doxxy.bot.versioning.github

import kotlinx.coroutines.runBlocking
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock
import kotlin.properties.ReadOnlyProperty
import kotlin.reflect.KProperty
import kotlin.time.Duration

class UpdateCountdown(duration: Duration) {
    private val interval: Long = duration.inWholeMilliseconds
    private var nextUpdate: Long = 0

    suspend fun onUpdate(block: suspend () -> Unit) {
        if (System.currentTimeMillis() > nextUpdate) {
            block()
            nextUpdate = System.currentTimeMillis() + interval
        }
    }
}

class UpdateCountdownDelegate<T : Any>(duration: Duration, private val updater: () -> T) : ReadOnlyProperty<Any, T> {
    private val lock = ReentrantLock()
    private val countdown = UpdateCountdown(duration)
    private lateinit var value: T

    override fun getValue(thisRef: Any, property: KProperty<*>): T {
        lock.withLock {
            runBlocking {
                countdown.onUpdate {
                    value = updater()
                }
            }
        }

        return value
    }
}